happi/audit.py

```diff
@@ -59,6 +61,54 @@ def check_name_match_id(result: SearchResult) -> None:
     assert id_field == name_field, f'id: {id_field} != name: {name_field}'
 
 
+def check_wait_connection(result: SearchResult) -> None:
+    """
+    Check if all the signals on the device can connect to their PV's
+
+    This can take a long time, consider filtering your query before running
+    """
+    dev = result.get()
+
+    assert (hasattr(dev, 'wait_for_connection') and
+            callable(getattr(dev, 'wait_for_connection'))), \
+           ('device has no wait_for_connection method, and is likely '
+            'not an ophyd v1 device')
+
+    try:
+        dev.wait_for_connection(timeout=5)
+    except TimeoutError as te:
+        # If we encounter a timeout error, gather some more detailed stats
+        sigs = list(sig.item for sig in dev.walk_signals())
+        conn_sigs = sum(sig.connected for sig in sigs)
+
+        raise ValueError(f'{conn_sigs}/{len(sigs)} signals connected. \n'
+                         f'original error: {te}')
+
+
+def check_args_kwargs_match(result: SearchResult) -> None:
+    """
+    Check that arguments that request information from the container
+    have information in those entries.
+
+    i.e.: makes sure there is information in the entry named "extra"
+          if the kwargs = {'extra': '{{extra}}'}
+    """
+    # Happi fills in values when search result is created, must pull
+    # raw document from client level
+    cl = result.client
+    doc = cl.find_document(**{cl._id_key: result[cl._id_key]})
+    # pick out any jinja-like template
+    env = Environment(undefined=DebugUndefined)
```

Comment:
Neat - is this better or a more modern version compared to the previous implementation? https://github.com/pcdshub/happi/blob/0cb79b5e1f6bb06e1c03e17e22fd316b755ac447/happi/loader.py#L56 (Just curious about the differences, not attempting to hold up this PR)

Reply:
I struggle to remember (though it hasn't been that long...), but I think this undefined kwarg is needed to make the Environment keep track of all the undeclared variables.  The loader tackles each arg individually, and this audit function tries to grab all undeclared at once.

I'd expect the default `Undeclared` handler to work, but after some experimentation I found it didn't.  So here we are 🤷